### DISCUSSION TITLE
Updates to with_timeout

### DISCUSSION THREAD

**Discussion by jwprice100** [cocotb/cocotb#2520] [Proposals]
# Goals
The ```with_timeout``` trigger is generally very useful. However in the gitter chat there was a good discussion on how it might be made more useful. The intention of this page is to capture those thoughts before any implementation work begins. Three Major Changes:
1. Accept an undecorated python coroutine, and generate a Cocotb task from it. This is/has been addressed in https://github.com/cocotb/cocotb/pull/2494. This will not be further discussed and is here for completeness. 

2. Generalize the interface to accept a timeout trigger, while being backwards compatible with the existing interface. One motivation behind this is timeouts are often done in terms of ClockCycles.

3. Implement a Timeout context manager. This would allow the timeout functionality to be used without creating a coroutine dedicated to it. This is very convenient for users.

# Example Usage
Update with_timeout interface to accept either a timeout_trigger, or number/units as it does today. Either of the examples given below should work:
```python
with_timeout(coro, 100, 'ns')
with_timeout(coro, ClockCycles(dut.clock, 32))
```

Timeout context manager:
```python
with Timeout(Timer(100, 'ns')):
    await coro1()
    await coro2()
```

# Implementation Considerations
Changes to with_timeout do not have any known issues. The changes should be fully backwards compatible and generally improve usage. 

Implementing a timeout context manager is more tricky. The basic mechanics would be to create and fork a timeout coroutine that awaits the timeout trigger upon the ``__enter__`` method. Upon ``__exit__`` the timeout coroutine is then killed. In order to allow that ``SimTimeoutError`` exception it was suggested that the parent coroutine (the one invoking the context manager) could have it's pending triggers modified to be a ``NullTrigger`` with ``SimTimeoutError``. 

Questions:
1. As I understand, use ``_resume_coro_upon`` to alter the trigger of a given task. However if the parent coroutine was the only one waiting on a given trigger, then the trigger needs to be cleaned up.

**Comment by ktbarrett** [comment]
1 is a good idea, I will try to finish that out soon. 2 is also a good idea. A good use case:

```python
await with_timeout(RisingEdge(dut.done), ClockCycles(dut.clk, 100))
```

3 I'm not a fan of at all. It is "weird" for the scheduler. I know we found a way to go about it, but I think there are other syntaxes that don't violate the current simulator assumption that a coroutine is only ever waiting on one trigger (which is shared with asyncio AFAICT). We could introduce a short syntax for forking coroutines:

```python
@run
async def a():
    # do stuff...
    await stuff()
    # yep...

await a  # wait until task is finished
```

And extend it with timeouts:

```python
@run_with_timeout(10, 'ns')
async def a():
    # do stuff...
    await stuff()
    # yep...

await a
```

This is the kind of stuff that belongs in a utility library. Things that exist in the main repo are common-enough features that could be reimplemented in C++ at some point. `run` and `run_with_timeout` can be implemented as abstractions on top of `fork` and `with_timeout` and don't *need* to be the main repo.

**Reply by jwprice100** [reply]
Cool,  I'll look forward to 1 and 2 in a future version! 

I hear your points about not breaking scheduler assumptions. I'm a tad puzzled by your statement of violating the assumption of only ever waiting on one coroutines. What the proposed implementation was is _changing_ the coroutine a scheduler is waiting on. This is probably still a bit...gross and probably still violates reasonable assumptions. Just want to make sure we're on the same page.

Your alternative syntax is interesting, I'd have to think about that. You raise the idea of a utility library. Is such a library planned to be created/semi-supported by the cocotb maintainers, similar to cocotb_bus? Or is this best left for the community to pick up and run with it?
